src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -9222,6 +9232,9 @@ private ErrorPropertySymbol CreateErrorPropertySymbol(ImmutableArray<PropertySym
                     methodGroup.Free();
                 }
             }
+            // PROTOTYPE: We'll likely want to collect all methods from compatible extension types
+            //            (for member access only, not for simple name, like we do for extension methods).
```

Comment:
>not for simple name [](http://example.com/codeflow?start=51&length=19)

Why is that? I think we look for extension methods even when there is no qualifier. It doesn't make sense to look for members when we are looking for a type (this is what we do in changed code in Binder_Symbols.cs), but in an expression context, I think we should assume there is an implicit qualifier - ```this```. Even in a type only context, is there a reason to not look for types inside a type extending the enclosing type?

Reply:
> I think we look for extension methods even when there is no qualifier. 

I don't think so. Extension methods do not come into play in simple name scenarios. 

Here's an [example](https://sharplab.io/#v2:EYLgtghglgdgPgWAFAAEDMACFAmDBhDZAb2QzK0xQBYMBZACgEpClyMTW3yAXACygDOAOgYBGRgG5SXMmMkYA9AowBTAE5qA9mvwBlUQFYADKJAYAcpoyaAbuoA2miABMMAM20YwKvptcByWn8MbggAaxUBDFEMCDUAcwBXbxhuAWkyAF9kbKRkdCxRADYsXABRAA9uFRgBKE0YYgyClGKsGgY+QXwMAGMAGgxYbiHmIgxM8hygA):

```
public class C 
{
    public void M() 
    {
        this.M(1); // ok
        M(1); // error CS1501: No overload for method 'M' takes 1 arguments
    }
}

public static class Extension
{
   public static void M(this C c, int i) { }   
}
```

> is there a reason to not look for types inside a type extending the enclosing type?

This is a spec question. The [current proposed "simple name" rules](https://github.com/dotnet/csharplang/blob/main/proposals/extensions.md#simple-names) do not add any extension member lookup. Note that they already didn't include any extension method lookup.

But I'm seeing something wrong with the official spec:
1. the "simple name" section doesn't mention extension methods (matches implementation), 
2. but the "method invocations" section does (which is wrong for invocations on simple names)
3. and the "member access" section does (but incorrectly, since it doesn't mention waiting for overload resolution to fail to proceed to extension method lookup)

I'll chat with Mads to clarify and see how to adjust the extensions spec correspondingly.

Reply:
Clarified with Mads how the current spec works for extension methods:
There are two ways of getting at extension method invocation:
1. if a member access yields an empty set
2. if a member access yields a method group, but the "method invocation" rules filter all of them out (by applicability rules)

We confirmed that extension method invocation only comes into play for member access, although that's done in a weird way:
In the "extension method invocation" section, the applicable forms all involve a member access.

We will do the same for extension types. That means that I'm missing a section in the spec for "method invocation" (and possibly some other sections involving overload resolution like "element access" too).

Based on that, I would propose to keep the current PR scoped as-is, it implements "extension member lookup" (non-generic) and plugs that into the "member access" as expected.
I can do a follow-up/additive PR for the cases where regular lookup succeeds but overload resolution fails (and thus we need to look at extensions). 
I'll add tests to illustrate this behavior.

Reply:
Here's a proposed update to the spec to include extension member lookup as part of the "method invocation" and "indexer access" sections:
[https://github.com/dotnet/csharplang/pull/7179](https://github.com/dotnet/csharplang/pull/7179)